File: src/to process/Shared/widgets-library/basics/option.ts
import { List } from "immutable";

import { Fun, Unit } from "./fun";
import { Sum } from "./sum";

export type Option<a> = Sum<Unit, a>;
export const none = function <a>(): Option<a> {
  return Sum.CreateLeft<Unit, a>({});
};
export const some = function <a>(v: a): Option<a> {
  return Sum.CreateRight<Unit, a>(v);
};
export const mapO = <a, b>(f: Fun<a, b>, o: Option<a>): Option<b> =>
  o.kind == "r" ? some(f(o.v)) : none();
export const onSome: <a>(f: Fun<a, void>, o: Option<a>) => void = mapO;
export const fromOption = function <a>(def: a, o: Option<a>): a {
  return o.kind == "r" ? o.v : def;
};
export const listFromOption = <a>(x: Option<a>): List<a> =>
  x.kind == "l" ? List<a>() : List<a>().push(x.v);
export const arrayFromOption = <a>(x: Option<a>): Array<a> =>
  x.kind == "l" ? [] : [x.v];

File: backend/libraries/ballerina-core-go/option.go
package ballerina

type Option[a any] struct {
	Sum[Unit, a] // NOTE: struct embedding is needed to be able to access Sum's JSON methods
}

// type some[a any] struct { Option[a]; Value a; Kind string }
func Some[a any](value a) Option[a] {
	return Option[a]{Right[Unit](value)}
}

// type none[a any] struct { Option[a]; Kind string }
func None[a any]() Option[a] {
	return Option[a]{Left[Unit, a](DefaultUnit)}
}

func MatchOption[a any, c any](self Option[a], onSome func(a) c, onNone func() c) c {
	return Fold(self.Sum, func(_ Unit) c { return onNone() }, onSome)
}

func MapOption[a any, b any](self Option[a], f func(a) b) Option[b] {
	return Option[b]{BiMap(self.Sum, id, f)}
}

type DeltaOptionEffectsEnum string
const (
  OptionReplace DeltaOptionEffectsEnum = "OptionReplace" 
  OptionValue DeltaOptionEffectsEnum = "OptionValue" 
)
var AllDeltaOptionEffectsEnumCases = [...]DeltaOptionEffectsEnum{ OptionReplace, OptionValue, }

func DefaultDeltaOptionEffectsEnum() DeltaOptionEffectsEnum { return AllDeltaOptionEffectsEnumCases[0]; }

type DeltaOption[a any, deltaA any] struct{
	DeltaBase
  Discriminator DeltaOptionEffectsEnum
  Replace a
  Value deltaA
}
func NewDeltaOptionReplace[a any, deltaA any](value a) DeltaOption[a, deltaA] {
  return DeltaOption[a, deltaA] {
    Discriminator:OptionReplace,
    Replace:value,
 }
}
func NewDeltaOptionValue[a any, deltaA any](delta deltaA) DeltaOption[a, deltaA] {
  return DeltaOption[a, deltaA] {
    Discriminator:OptionValue,
    Value:delta,
 }
}


func MatchDeltaOption[a any, deltaA any, Result any](
  onReplace func(a) (Result, error),
  onValue func(deltaA) (Result, error),
) func (DeltaOption[a, deltaA]) (Result, error) {
  return func (delta DeltaOption[a, deltaA]) (Result,error) {
    var result Result
    switch delta.Discriminator {
      case "OptionReplace":
        return onReplace(delta.Replace)
      case "OptionValue":
        return onValue(delta.Value)
    }
    return result, NewInvalidDiscriminatorError(string(delta.Discriminator), "DeltaOption")
  }
}
